mm2src/mm2_main/src/lp_swap/swap_watcher.rs

```diff
@@ -226,55 +225,95 @@ impl State for WaitForTakerPaymentSpend {
-                return Self::change_state(RefundTakerPayment {});
-            },
+        const TAKER_PAYMENT_SPEND_SEARCH_INTERVAL_SLOW: f64 = 300.;
+        let payment_search_interval = match std::env::var("USE_TEST_SEARCH_INTERVAL") {
+            Ok(_) => TAKER_PAYMENT_SPEND_SEARCH_INTERVAL,
+            Err(_) => TAKER_PAYMENT_SPEND_SEARCH_INTERVAL_SLOW,
         };
-
-        let tx_hash = tx.tx_hash();
-        info!("Taker payment spend tx {:02x}", tx_hash);
-        let tx_ident = TransactionIdentifier {
-            tx_hex: BytesJson::from(tx.tx_hex()),
-            tx_hash,
+        let wait_until = match std::env::var("SKIP_WAIT_FOR_TAKER_PAYMENT") {
+            Ok(_) => 0,
+            Err(_) => watcher_ctx.data.swap_started_at + (4 * watcher_ctx.data.lock_duration / 3),
+        };
+        let search_input = WatcherSearchForSwapTxSpendInput {
+            time_lock: (watcher_ctx.data.swap_started_at + watcher_ctx.data.lock_duration) as u32,
+            taker_pub: &watcher_ctx.verified_pub,
+            maker_pub: &watcher_ctx.data.maker_pub,
+            secret_hash: &watcher_ctx.data.secret_hash,
+            tx: &self.taker_payment_hex,
+            search_from_block: watcher_ctx.data.taker_coin_start_block,
+            wait_until,
+            check_every: payment_search_interval,
         };
 
-        let secret = match watcher_ctx
+        let f = watcher_ctx
             .taker_coin
-            .extract_secret(&watcher_ctx.data.secret_hash, &tx_ident.tx_hex.0)
-            .await
-        {
-            Ok(bytes) => H256Json::from(bytes.as_slice()),
+            .watcher_search_for_swap_tx_spend(search_input)
+            .await;
+
+        match f {
+            Ok(FoundSwapTxSpend::Spent(tx)) => {
+                let now = now_ms() / 1000;
+                if now < watcher_ctx.data.swap_started_at + watcher_ctx.data.lock_duration {
+                    let wait_until = match std::env::var("SKIP_WAIT_FOR_MAKER_PAYMENT") {
+                        Ok(_) => 0,
+                        Err(_) => watcher_ctx.data.swap_started_at + watcher_ctx.data.lock_duration,
+                    };
+
+                    let maker_payment_hex_fut = watcher_ctx
+                        .maker_coin
+                        .get_tx_hex_by_hash(watcher_ctx.data.maker_payment_hash.clone());
+                    let maker_payment_hex = match maker_payment_hex_fut.compat().await {
+                        Ok(tx_res) => tx_res.tx_hex.into_vec(),
+                        Err(err) => {
+                            return Self::change_state(Stopped::from_reason(StopReason::Error(
+                                WatcherError::MakerPaymentCouldNotBeFound(err.to_string()).into(),
+                            )))
+                        },
+                    };
+
+                    let f = watcher_ctx.maker_coin.wait_for_htlc_tx_spend(
+                        &maker_payment_hex,
+                        &[],
```

Comment:
Providing empty secret might lead to a bug for some coins. Please consider using `&watcher_ctx.data.secret_hash` here.